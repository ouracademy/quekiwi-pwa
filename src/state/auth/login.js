import { getStandardRequestFor } from "../standard-request"
import { authStandardOptions } from "./standard-options"
import { ajax } from "rxjs/ajax"

const { actionCreators, reducer, epicFrom } = getStandardRequestFor(
  "LOGIN",
  authStandardOptions
)

export const [login, loginSuccessFully] = actionCreators

export { reducer }

export const loginEpic = epicFrom(payload =>
  ajax.post("http://localhost:3000/auth/login", payload)
)
