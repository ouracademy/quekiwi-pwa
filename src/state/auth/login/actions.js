export const LOGIN_REQUESTED = "LOGIN"
export const LOGIN_SUCCESSFULLY = "LOGIN_SUCCESSFULLY"
export const LOGIN_FAILED = "LOGIN_FAILED"

export const login = credentials => ({
  type: LOGIN_REQUESTED,
  payload: credentials,
})

export const loginSuccessFully = ({ response }) => ({
  type: LOGIN_SUCCESSFULLY,
  payload: {
    token: response.token,
  },
})
