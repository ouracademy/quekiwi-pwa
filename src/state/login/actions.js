export const LOGIN = "LOGIN"
export const LOGIN_SUCCESSFULLY = "LOGIN_SUCCESSFULLY"
export const LOGIN_FAILED = "LOGIN_FAILED"

export const login = credentials => ({
  type: LOGIN,
  payload: credentials,
})

export const loginSuccessFully = ({ response }) => ({
  type: LOGIN_SUCCESSFULLY,
  payload: {
    token: response.token,
  },
})
