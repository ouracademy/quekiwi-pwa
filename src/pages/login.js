import React, { useState } from "react"
import { Link } from "gatsby"

import SEO from "../components/seo"
import { Box, Button, Form, FormField, Heading } from "grommet"
import { FormClock } from "grommet-icons"
import { connect } from "react-redux"
import { login } from "../state/actions"

const LoginPage = ({ dispatch }) => (
  <Box height="100vh" width="100vw" align="center" justify="center" pad="small">
    <SEO title="Login" keywords={[`login`, `quekiwi`, `libros`]} />
    <Heading level={4}>Ingresa a Quekiwi</Heading>
    <Box width="medium">
      <LoginForm dispatch={dispatch} />
      <Suggestion />
    </Box>
  </Box>
)

const LoginForm = ({ dispatch }) => {
  const [isSubmitting, setIsSubmitting] = useState(false)
  const submit = ({ value }) => {
    setIsSubmitting(true)
    dispatch(login(value))
    setIsSubmitting(false)
  }
  const loginData = { name: "", email: "" }

  return (
    <Form onSubmit={submit} value={loginData}>
      <FormField label="Correo" name="email" required type="email" />
      <FormField label="ContraseÃ±a" required name="password" type="password" />
      <Box direction="row" justify="center" margin={{ top: "medium" }}>
        {isSubmitting ? (
          <FormClock />
        ) : (
          <Button type="submit" fill label="Ingresar" primary />
        )}
      </Box>
    </Form>
  )
}

export const Suggestion = () => (
  <p>
    <Link to="/sign-up">Â¿AÃºn no tengo cuenta?</Link> Pero deseo registrarme{" "}
    <span role="img" aria-label="feliz">
      ðŸ™‚
    </span>
  </p>
)

export default connect()(LoginPage)
