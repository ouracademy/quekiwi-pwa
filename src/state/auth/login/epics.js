import { LOGIN_REQUESTED, loginFailed, loginSuccessFully } from "./actions"
import { standardEpic } from "../../standard-request/epic"

export const loginEpic = standardEpic(
  LOGIN_REQUESTED,
  loginSuccessFully,
  loginFailed
)("http://localhost:3000/auth/login")
