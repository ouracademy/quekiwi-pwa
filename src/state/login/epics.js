import { ofType } from "redux-observable"
import { ajax } from "rxjs/ajax"
import { mergeMap, map, catchError } from "rxjs/operators"
import { of } from "rxjs"
import { LOGIN, LOGIN_FAILED, loginSuccessFully } from "./actions"

export const loginEpic = action$ =>
  action$.pipe(
    ofType(LOGIN),
    mergeMap(action =>
      ajax.post("http://localhost:3000/auth/login", action.payload).pipe(
        map(loginSuccessFully),
        catchError(error =>
          of({
            type: LOGIN_FAILED,
            payload: error.response.message || "Ups, ocurri√≥ un error",
          })
        )
      )
    )
  )
