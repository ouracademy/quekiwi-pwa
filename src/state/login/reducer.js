import { LOGIN, LOGIN_SUCCESSFULLY, LOGIN_FAILED } from "./actions"

export const loginReducer = (
  state = { logged: false, loading: false, error: null, token: null },
  action
) => {
  const { type, payload } = action
  switch (type) {
    case LOGIN:
      return { loading: true, error: null }
    case LOGIN_SUCCESSFULLY:
      return { loading: false, error: null, token: payload.token, logged: true }
    case LOGIN_FAILED:
      return { loading: false, error: payload }
    default:
      return state
  }
}
