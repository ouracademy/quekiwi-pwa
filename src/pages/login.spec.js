import React from "react"
import renderer from "react-test-renderer"
import { Suggestion } from "./login"

describe("Suggestion", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Suggestion />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

test("1 + 1 = 2", () => {
  expect(1 + 1).toBe(2)
})
