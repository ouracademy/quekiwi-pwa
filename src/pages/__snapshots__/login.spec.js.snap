// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suggestion renders correctly 1`] = `
<p>
  <a
    href="/sign-up"
  >
    Â¿AÃºn no tengo cuenta?
  </a>
   Pero deseo registrarme
   
  <span
    aria-label="feliz"
    role="img"
  >
    ðŸ™‚
  </span>
</p>
`;
